akka-stream/src/main/scala/akka/stream/remote/scaladsl/SinkRef.scala

```diff
@@ -0,0 +1,199 @@
+        log.warning("Demanding until {}, from {}", StreamRefs.CumulativeDemand(localCumulativeDemand), remotePartner)
+      }
+
+      def runningReceive(activeSender: ActorRef): ((ActorRef, Any)) ⇒ Unit = {
+        case (`activeSender`, msg) ⇒
+          log.warning("RECEIVED: {}", msg)
+      }
+
+      lazy val initialReceive: ((ActorRef, Any)) ⇒ Unit = {
+        case (sender, StreamRefs.SequencedOnNext(seqNr, payload: T)) ⇒
+          log.warning("Received seq {} from {}", StreamRefs.SequencedOnNext(seqNr, payload: T), sender)
+          remotePartner = sender
+          self.watch(sender) // FIXME?
+
+          triggerDemand()
+          push(out, payload) // TODO only if allowed to push
+          getStageActor(runningReceive(sender)) // become running
+
+      }
+
+      setHandler(out, this)
+    }
+    (logic, promise.future) // FIXME we'd want to expose just the ref!
+  }
+}
+//
+//object SinkRefReceiverActor {
+//  def props(stageRef: ActorRef): Props =
+//    Props(new SinkRefReceiverActor(it ⇒ stageRef ! it))
+//}
+//class SinkRefReceiverActor[T](pushLocal: T ⇒ Unit) extends Actor with ActorLogging {
+//  /**
+//   * Scala API: This defines the initial actor behavior, it must return a partial function
+//   * with the actor logic.
+//   */
+//  override def receive: Receive = {
+//    case StreamRefs.SequencedOnNext(seqNr, payload: T) ⇒
+//      context.watch(sender()) // FIXME, watch just once
+//      sender ! StreamRefs.CumulativeDemand(16) // TODO configure demand management
+//
+//      pushLocal(payload) // TODO only if allowed to push
+//  }
+//}
+
+/** Do not create this instance directly, but use `SinkRef` factories, to run/setup its targetRef */
+final class SinkRef[In] private[akka] (private[akka] val targetRef: ActorRef) extends GraphStage[SinkShape[In]]
+  with Serializable { stage ⇒
+  import akka.stream.remote.StreamRefs._
+
+  @transient val in = Inlet[In](s"SinkRef($targetRef).in")
+  override def shape: SinkShape[In] = SinkShape.of(in)
+
+  override def createLogic(inheritedAttributes: Attributes) = new GraphStageLogic(shape) with StageLogging with InHandler {
+
+    // we assume that there is at least SOME buffer space
+    private var knownRemoteDemand = 4L // FIXME figure it out, OR we need a handshake
+
+    // FIXME this one will be sent over remoting so we have to be able to make that work
+    @transient var self: GraphStageLogic.StageActor = _
+    implicit def selfRef = self.ref
+
+    override def preStart(): Unit = {
+
+      self = getStageActor {
+        case (_, Terminated(`targetRef`)) ⇒
+          failStage(failException())
+
+        case (sender, CumulativeDemand(d)) ⇒
+          validatePartnerRef(sender)
+
+          log.warning("Received demand {} from {}", CumulativeDemand(d), targetRef)
+          if (knownRemoteDemand == 0) {
+            knownRemoteDemand = d
+            tryPull()
+          } else tryPull()
```

Comment:
Why are we throwing the demand away here?

Reply:
Bug, thanks! Should always update if the incoming is > current

Reply:
// Need to move this entire class to the cumulative demand